servers/physics_3d/godot_physics_server_3d.cpp

```diff
@@ -175,6 +175,44 @@ bool GodotPhysicsServer3D::space_is_active(RID p_space) const {
 	return active_spaces.has(space);
 }
 
+void GodotPhysicsServer3D::space_step(RID p_space, real_t p_delta) {
+	GodotSpace3D *space = space_owner.get_or_null(p_space);
+	ERR_FAIL_NULL(space);
+
+	// TODO:
+	// May be let pending_shape_update_list as a member of GodotSpaces3D and update shapes by themselves.
+	// To avoid effecting Spces which are handled by developer (for lockstep/rollback netcode, it is particularly sensitive).
+	// If it is unnecessary, call _update_shapes() directly.
+	SelfList<GodotCollisionObject3D> *collision_object_self = pending_shape_update_list.first();
+	while (collision_object_self) {
+		if (collision_object_self->self()->get_space() == space) {
+			collision_object_self->self()->_shape_changed();
+
+			SelfList<GodotCollisionObject3D> *to_remove = collision_object_self;
+			collision_object_self = collision_object_self->next();
+
+			pending_shape_update_list.remove(to_remove);
+		} else {
+			collision_object_self = collision_object_self->next();
+		}
+	}
+
+	stepper->step(space, p_delta);
+}
```

Comment:
I'm _far_ from an expert on Godot's internals, but after digging around for a while I've come to this conclusion: Just call `_update_shapes()` directly. It looks like `_update_shapes()` just updates the internal physics world with any changes to the shapes of the bodies. So for instance, if a user adds a new CollisionShape3D node, it won't immediately create the real physics shape internally. The internal representation gets updated when `_update_shapes()` is called, which happens before anything that relies on the shapes being updated. For example, if you call `GodotPhysicsServer3D::body_apply_impulse`, then `_update_shapes()` will also be called because applying an impulse relies on the shapes being up to date.

I think ideally, `_update_shapes()` should actually be changed to take a parameter `_update_shapes(RID p_space)` so that only spaces that actually need updating get updated, but I think that should be only a minor performance optimisation (unless people have many spaces and are frequently updating shapes in a weird way within a single frame, like: changeShapeInSpace1, applyImpulse, changeShapeInSpace2, applyImpulse, changeShapeInSpace1, applyImpulse, etc.) and can probably be done in a separate PR.

So, I think this should just be:

```cpp
void GodotPhysicsServer3D::space_step(RID p_space, real_t p_delta) {
	GodotSpace3D *space = space_owner.get_or_null(p_space);
	ERR_FAIL_NULL(space);

	_update_shapes();
	stepper->step(space, p_delta);
}
```

But this is only the result of an hour of digging and I could be wrong. Please correct me if I am.

Reply:
Also not an expert, but I think the main question is "is it safe to apply pending shape updates from **all** spaces, any time a single space is stepped?" It does seem a little wrong for a space-specific method to affect other spaces, although I doubt it would cause any real problems for anyone.

I agree with the idea that `_update_shapes()` should probably be space-specific, and the overall `step()` method could call it for each active space

Reply:
I believe the answer to that question is "yes it's safe". I think the only real consideration is that it's a (slight) performance issue in some very niche and specific cases. I'm not sure what the convention is here but I would assume we'd probably want to fix that in a separate PR so that we're only changing one thing at a time, especially since this PR has zero impact on existing projects since it only adds new functions.

On a related note, it's also kind of weird that functions like `GodotPhysicsServer3D::body_apply_impulse` call `_update_shapes()`, which applies pending shape updates for _all_ bodies, not just the relevant ones to the function.

Reply:
I agree "_update_shapes(RID p_space)", but not for performance.
This pr is for deterministic resimulation, it should what it should do only. We can't assume that user logic are not depend on the order of updating shapes. 

And  "_update_shapes(RID p_space)" is similar to current implement, so I prefer to keep it, unless there is need to reuse this function.